downstairs/src/lib.rs

```diff
@@ -1499,106 +1554,231 @@ impl Downstairs {
+                // information, as that will not be valid any longer.
+                //
+                // TODO: Really work through this error case
+                if work.active.keys().len() > 0 {
                     println!(
-                        "Error while signaling to {:?} thread: {:?}",
-                        old_upstairs.0, e,
+                        "Crucible Downstairs promoting {:?} to active, \
+                        discarding {} jobs",
+                        upstairs_connection,
+                        work.active.keys().len()
                     );
                 }
-            }
-        }
 
-        self.active_upstairs = Some((upstairs_connection, tx));
+                // In the future, we may decide there is some way to continue
+                // working on outstanding jobs, or a way to merge. But for now,
+                // we just throw out what we have and let the upstairs resend
+                // anything to us that it did not get an ACK for.
+                work.clear();
+            } else {
+                // There is no current session for this Upstairs UUID.
+            }
 
-        /*
-         * Note: in the future, differentiate between new upstairs connecting
-         * vs same upstairs reconnecting here.
-         *
-         * Clear out active jobs, the last flush, and completed information,
-         * as that will not be valid any longer.
-         *
-         * TODO: Really work through this error case
-         */
-        if work.active.keys().len() > 0 {
-            println!(
-                "Crucible Downstairs promoting {:?} to active, \
-                discarding {} jobs",
-                upstairs_connection,
-                work.active.keys().len()
+            // Insert a new session, overwritting the previous entry if the
+            // Upstairs UUID has an entry already.
+            self.active_upstairs.insert(
+                upstairs_connection.upstairs_id,
+                ActiveUpstairs {
+                    upstairs_connection,
+                    work: Mutex::new(Work::new()),
+                    terminate_sender: tx,
+                },
             );
 
-            /*
-             * In the future, we may decide there is some way to continue
-             * working on outstanding jobs, or a way to merge. But for now,
-             * we just throw out what we have and let the upstairs resend
-             * anything to us that it did not get an ACK for.
-             */
-            work.active = HashMap::new();
-        }
+            Ok(())
+        } else {
+            // Only one active read-write session is allowed. Kick out the
+            // currently active Upstairs session if one exists.
+            let currently_active_upstairs_uuids: Vec<Uuid> =
+                self.active_upstairs.keys().copied().collect();
+
+            match currently_active_upstairs_uuids.len() {
```

Comment:
Similar locking question here, thought it seems like we are okay.  The action of getting
currently_active_upstairs_uuids, that is atomic?  Meaning, could two connections both call that `self.active_upstairs.keys().copied().collect();` and come back with `0`?

Actually, I think it could, as the time of check is different that the time of add below, so unless we
lock around this, I suspect it's possible for two connections to sneak in here.

Reply:
I think we're ok here - `promote_to_active` needs a `&mut self`, meaning an exclusive reference, meaning I don't think two calls of this can race.